### ISSUE TITLE
[GPT-OSS-120B,b=1,WHGLX] SamplingDeviceOperation hang during decode trace replay on g04glx01

### ISSUE THREAD

**Issue by handrewsTT** [tenstorrent/tt-metal#45934]
## TL;DR

GPT-OSS-120B tier-1 e2e on a WH Galaxy hung in `test_gpt_oss_demo[wormhole_b0-4x8-prefill_128]` (b=1, isl=128) on **g04glx01**. tt-triage pins the stuck op to **`SamplingDeviceOperation` (op 11588, trace id 2)** running on **Device 21, core (4,1) / NOC (6,2)** — preceded by `ManualSeedDeviceOperation` (op 11587). All four RISCs on that core report `PC was not in range of any provided ELF files`, while cq_prefetch / cq_dispatch on the dispatcher are wedged on `fetch_q_get_cmds` / `CBReader::acquire_pages`.

The TT_THROW signature is:

```
TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable
TT_THROW: Device 0: Timeout (5000 ms) waiting for physical cores to finish: 19-27, 18-27.
```

This is likely the same family of failure as #45327 (same job, same test, same runner), but tt-triage in this run isolates **the sampling op specifically** which the earlier issue did not pin down — so opening a fresh issue with the new evidence and a sharper repro target. Will dupe-close into #45327 if root-cause turns out to be identical.

## Failing run

- Workflow: `(Tier 1) Models E2E [scheduled]` — `models-e2e-tests / GPT-OSS 120B e2e tests [wh_galaxy_perf]`
- Run: https://github.com/tenstorrent/tt-metal/actions/runs/26861720371
- Job: https://github.com/tenstorrent/tt-metal/actions/runs/26861720371/job/79217498940
- Branch: `main` @ `8d1c028eeae15a5780b4e4a869851939a71e2412`
- Runner: `g04glx01` (same host that #45327 keeps hitting)
- Started: 2026-06-03T03:33:32Z → first TT_THROW at 03:38:43Z (~5 min into the test)

```
pytest models/demos/gpt_oss/demo/text_demo.py --timeout 1500 --skip-model-load
```
`HF_MODEL=openai/gpt-oss-120b`, `TT_MM_THROTTLE_PERF=2`, `TT_METAL_OPERATION_TIMEOUT_SECONDS=5`.

All 11 subsequent parametrizations (`prefill_1k`, `prefill_4k`, …, `batch128`) error out immediately with `SubDeviceManagerTracker is not initialized on MeshDevice 0` because the device is unrecoverable after the first hang — same cascade as #45327.

### `dump_running_operations.py` — last op in flight

```
┃ Op Id ┃ Op Name                               ┃ Op Params      ┃ Prev Op Id ┃ Prev Op Name              ┃ Device Cnt ┃ Core Cnt ┃ Devices ┃ Cores        ┃
│ 11588 │ SamplingDeviceOperation (trace id: 2) │ Tensor[0..5]   │ 11587      │ ManualSeedDeviceOperation │ 1          │ 1        │ 21      │ 21:6-2 (4,1) │
```

Input shapes to the sampling op (truncated):

- `Tensor[0]`: `[1, 1, 32, 256]`, bfloat16, INTERLEAVED/DRAM (the logits)
- `Tensor[1]`: `[1, 1, 32, 256]`, uint32 (top-k indices, from a prior topk)
- `Tensor[2..4]`: `[32]` (per-token K/p/temperature)
- `Tensor[5]`: `[1, 1, 1, 32]`, uint32 (manual seed tensor — output of op 11587)

### `dump_aggregated_callstacks.py` — what the stuck core was doing

On **Device 21 / (4,1) / 6-2**, all four RISCs were running `sampling` / `writer_interleaved` / `reader_values_indices_tensor` kernels but with `PC was not in range of any provided ELF files` — i.e. PC has wandered out of the kernel's `.text` range (could be a stack overflow, return-address corruption, or jump into garbage from a malformed indirect call).

Meanwhile on the dispatcher side:

```
cq_prefetch     brisc   fetch_q_get_cmds<0> ()              at cq_prefetch.cpp 763:30
cq_dispatch     brisc   CBReader::acquire_pages ()          at cq_common.hpp   459:52
                        wait_for_available_data_and_release_old_pages
                        kernel_main ()                       at cq_dispatch.cpp 1537:102
```

— classic "dispatcher is blocked because a worker hasn't signalled completion".

### Other checks

| Check | Result |
|---|---|
| `check_arc.py` | pass — all 32 chips heart-beating (devs 26/28/29/31 at 9.48 Hz vs 9.98 Hz for the rest, but within nominal) |
| `check_eth_status.py` | pass |
| `check_noc_locations.py` | pass |
| `check_noc_status.py` | pass |
| `check_cb_inactive.py` | pass |
| `check_core_magic.py` | pass |
| `check_binary_integrity.py` | pass |
| `check_broken_components.py` | pass |
| Firmware bundle | 19.8.1 (newer than tested 19.5.0 — already flagged as a UMD warning) |

So no ARC / eth / NOC fabric anomaly is detected. The hang is **at the kernel level on a single sampling worker core**, not a system-wide fabric/eth event.

## Possible root causes

1. **PR #45747 — "Declare device sampling capability on model side." (merged 2026-06-03 02:25 UTC, ~1 hour before this job started).** This PR touched `models/tt_transformers/tt/generator.py` and `generator_vllm.py` to declare device-side sampling capability. GPT-OSS uses tt_transformers. The hang is in `SamplingDeviceOperation` exactly the path this PR newly enables (or changes the launch conditions for) — **prime suspect**. Worth checking whether this run is the first nightly tier-1 e2e since that PR landed.

2. **Sampling kernel has a real bug at b=1**. The sampling kernel source (`ttnn/cpp/ttnn/operations/reduction/sampling/device/kernels/compute/sampling.cpp`) shows deprecation-warning churn (`acquire_dst`/`release_dst` vs `tile_regs_*`) suggesting recent edits, and the kernel runs on a **single core** for b=1 — so any per-tile branch that loops on a degenerate input (e.g. `K=1`, or all-equal logits leading to a normalize-divide-by-zero path) would manifest exactly as "kernel runs forever, PC drifts out of text". A `nan`/`-inf` in the logits after a freshly-loaded seed tensor could plausibly drive this.

3. **Trace-replay vs eager divergence around `ManualSeed → Sampling`.** Op 11587 is a `ManualSeedDeviceOperation` writing to a `[1,1,1,32]` uint32 seed tensor that op 11588 consumes (as `Tensor[5]`). If the trace captured the seed-tensor address but the replay path re-allocates that tensor at a different DRAM location (or fails to push to the consumer CB on the first replay iteration), Sampling can spin in `cb_wait_front` on a CB that will never be filled. This matches the "dispatcher idle in `acquire_pages`" symptom on the dispatch side.

4. **PC-out-of-ELF could indicate firmware/UMD mismatch on this specific runner.** The UMD reports `Firmware bundle version 19.8.1 on the system is newer than the latest fully tested version 19.5.0 for wormhole_b0`. If `g04glx01` is on a non-uniform FW (relative to other hosts), a kernel might be miscompiled / mislinked against the wrong runtime symbols. #45517 already implicates kernel `.text` corruption on a galaxy host — could be a related class of failure.

5. **Same-host clustering.** This is **at least the second prefill_128 hang on `g04glx01`** in recent weeks (cf. #45327). If the stress matrix in #45327 confirms the failure rate is much higher on g04glx01 than on g04glx03 / g05glx04 / g14glx03 / g14glx04, the hang has a hardware-conditional component that the sampling kernel is the *trigger* for but not the *cause* of.

## Recommended next steps

1. **Bisect against PR #45747.** Re-run the same job on the parent commit `62aebe1ce86` (one before #45747). If it passes consistently on g04glx01, blame is on #45747; if it still hangs, this is the pre-existing #45327 family.
2. **Re-run with `TT_METAL_WATCHER=1` enabled on g04glx01.** The watcher log will capture the in-flight NOC transaction that the sampling kernel was wedged on (`cb_wait_front`, `noc_async_read`, or stack overflow detection) and tell us in one line whether this is kernel-internal vs CB-handshake.
3. **Pin to g04glx01 and stress-loop just `prefill_128`.** Build on the per-host SKU plumbing from `divanovic/gpt-oss-120b-prefill-128-stress` (already used in #45327) but run 20× / job rather than 5× to characterize the new failure rate after #45747.
4. **Inspect sampling kernel for b=1 path.** Audit `ttnn/cpp/ttnn/operations/reduction/sampling/device/kernels/compute/sampling.cpp` for any branch where the inner loop bound is derived from `batch` (or a `K` / `temperature` per-batch field) and could loop indefinitely for a degenerate b=1 input. The `acquire_dst`/`release_dst` deprecation noise suggests this file has had recent edits — `git log -p` on the last 4 weeks should be short.
5. **Compare g04glx01 firmware bundle against other galaxies.** If FW differs between hosts, normalize it before treating this as a pure software bug.
6. **Cross-link with #45327 (#45052 for the BHQB2 first-iter hang and #45517 for the kernel `.text` corruption story).** If the bisect lands on #45747, this is a new bug; otherwise dupe-close into the relevant existing thread.

Owner: TBD. Suggest pulling in the sampling-op author(s) and the #45747 PR author once initial triage above narrows it down.

**Comment by ifabijanic-tt** [comment]
## Could not reproduce on wh-glx6u-04 and wh-glx6u-08 - 276 `prefill_128` iterations, 0 SamplingDeviceOperation failures

Attempted to reproduce on a different Galaxy 6U hosts using the issue's exact invocation and env. No sampling hang. Two caveats about how far this generalises are at the bottom.

### Environment

| | |
|---|---|
| Host | wh-glx6u-04 and wh-glx6u-08 (not g04glx01) |
| Branch / SHA | main @ 6aecbb35477 |
| Firmware bundle | **19.2.0** (the failing host reported 19.8.1) |
| UMD | 4ed96fb3 (v0.9.5-194) |
| Driver | TT-KMD 2.5.1-pre |
| Build | clean rebuild (_build_metal.sh --clean_ then _build_metal.sh_), 1389/1389 targets, no errors |
| ttnn weight cache | wiped and regenerated from HF weights against this build |
| Mesh | 4×8 |

**The prime suspect is present in this build.** PR #45747 ("Declare device sampling capability on model side", commit `61d4e1a1f07`) is an ancestor of the tested HEAD, so the code path H1 points at was active.

### What was run

The issue's invocation and env, verbatim, plus `TT_METAL_WATCHER=1` as recommended:

```
TT_MM_THROTTLE_PERF=2  TT_METAL_OPERATION_TIMEOUT_SECONDS=5  TT_METAL_WATCHER=1
HF_MODEL=openai/gpt-oss-120b
pytest models/demos/gpt_oss/demo/text_demo.py --timeout 1500 --skip-model-load \
  -k '4x8 and prefill_128 and not prefill_128k'
```

Env vars were verified present in the actual pytest process via `/proc/<pid>/environ`, not merely exported by the parent shell — `TT_METAL_OPERATION_TIMEOUT_SECONDS=5` matters here, since it is what converts a silent spin into the reported error within 5s.

Structured as short rounds (5 warm iterations per process, `--count=5 -x`) rather than a few long runs, because the reported hang lands  around 5 minutes in, i.e. in the first iterations after startup, so the budget buys many fresh process starts. Each round ran around 8-10 minutes, which is comfortably past that window.

### Result

**56 rounds / 276 `prefill_128` iterations. 55 rounds clean. Zero SamplingDeviceOperation failures.**

One round (9) failed on machine _wh-glx6u-04_, but **not with this bug**:

```
Watcher detected that active eth link on virtual core 24-17 (noc0 core: e8-6) went down after training.
TT_THROW: Device 7 acteth core(x=0,y=10) virtual(x=24,y=17)
  While running kernels: erisc: tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_router.cpp
```

That is the known intermittent ETH-link issue, not this one. Worth flagging explicitly because it surfaced the **same** top-level string quoted in this issue:

> `TT_THROW: TIMEOUT: device timeout, potential hang detected, the device is unrecoverable`

That message is generic,  it fires for any op exceeding `TT_METAL_OPERATION_TIMEOUT_SECONDS`, whatever the underlying cause. In this failure `SamplingDeviceOperation` appears **zero** times in the log; the in-flight kernel was the fabric erisc router on Device 7, not sampling on Device 21. **Anyone triaging by that string alone risks conflating an ETH-link drop with this sampling hang.**

### Limits of this negative result

Two gaps that couldn't be closed on these machines, both material:

1. **Firmware.** These hosts runs bundle **19.2.0**; the failing host ran **19.8.1**. H4 (UMD/kernel symbol mismatch from the newer bundle) is therefore **untested**, and it is the one hypothesis this run cannot speak to at all.
2. **Code age.** The tested HEAD is **2,080 commits newer** than the reported SHA `8d1c028eeae` (Jul 29 vs Jun 2). A negative here means "not reproducible on current main", not "the June failure was not real". If the goal is to reproduce the original report specifically, it needs a build at that SHA.

I also (probably) did not match `g04glx01`'s thermal/load state or its surrounding CI job ordering.

Potential next steps could be these:

1. **Run this on specifically `g04glx01`.** All three recent GPT-OSS P1 hangs (#45327, #45517, #45934) were reported on `g04glx01`/`g14glx03`, and none of the three reproduce here. Host-specificity/bad state is almost 100% the issue here.
2. **Check firmware bundle consistency across the Galaxy fleet** (H4). If `g04glx01` is the only host on 19.8.1, that is a cheap and high-value thing to confirm or rule out before deeper kernel work. Small chance that this is the issue, but maybe could be investigated.

Otherwise we can tell that this version of tt-metal code doesn't trigger this issue and this issue is not relevant any more.

**Comment by mtairum** [comment]
Due to the validation runs in July 30 being all green, I'm closing this issue.
Reopen if it starts showing up again.
